docs/src/main/sphinx/develop/tests.md

```diff
@@ -0,0 +1,424 @@
+  for loops or multiple data provider parameters
+
+### Avoid writing stateful test classes
+
+Stateful tests can lead to issues from on one test leaking into other tests,
+especially when test runs are parallelized. As a result debugging and
+troubleshooting test failures and maintenance of the tests is more difficult. If
+possible these stateful test classes should be avoided.
+
+### Do not try to manage memory
+
+JUnit and the JVM take care of test life cycle and memory management. Avoid
+manual steps such as nulling out fields in `@After` methods to “free memory”. It
+is safe to assign memory intensive objects to final fields, as the class is
+automatically dereferenced after the test run.
+
+### Use simple resource initialization
+
+Prefer resource initialization in constructors and tear them down in `@After`
+methods if necessary. This approach, combined with not nulling fields, allows
+the fields to be final and behave like any `Closeable` class in normal Java code
+Consider using the Guava `Closer` class to simplify cleanup.
+
+### Keep test setup and teardown simple
+
+Avoid the `@Before`/`@After` each test method style of setup and teardown.
+
+* Prefer try-with-resources if natural
+* If necessary, use a shared initialization or cleanup method that is explicitly
+  called.
+* If you have a test that benefits from @Before/After methods, discuss the
+  approach with the maintainers to develop a solution and improve guidance.
+
+### Ensure testability of new plugin and connector features
+
+New plugin/connector features should be testable using one of the testing
+plugins (e.g., memory or null). There are existing features only tested in
+plugins in Hive, and over time we expect coverage using the testing plugins
+
+### Keep focus on plugin and connector tests
+
+For plugins and specifically connector plugins, focus on the code unique to the
+plugin. Do not add tests for core engine features. Plugins should be focused on
+the correctness of the SPI implementation, and compatibility with external
+systems.
+
+### Avoid flaky tests
+
+Existing flaky tests using the legacy TestNG library can be marked with the
+`@Flaky` annotation temporarily to improve CI reliability until a fix is
+implemented:
+
+* Ideally the fix is to make the test reliable.
+* Rewrite the test to not rely on flakey infrastructure, including the practice
+  to avoid HDFS.
+* If necessary, add explicit retries, but be cognizant of resource usage.
+
+After a certain time period, if the test hasn’t been fixed, it should be
+removed.
+
+New tests with the `@Flaky` annotation can no be introduced, since new tests
+must use JUnit. Rewrite the test to be stable or avoid the test altogether.
+
+### Avoid disabling tests
+
+Prefer to remove a test instead of disabling it. Test code is maintained and
+updated as the codebase changes, and inactive tests just waste time and effort.
+
+Disabled tests can be removed at any time.
+
+### Avoid using `Assumptions.abort()`
```

Comment:
Is there a preferred solution?

Reply:
> what if CI build just stop as soon as a test fails?

That might be useful for CI but very limiting for engineers because then you need more round-trips to find all the failures and fix them.

The question is basically: how to you ignore a test conditionally instead of marking it as failed. e.g. cannot test merge queries against JDBC connectors but can in others.